_posts/2018-11-02-cpp-array.md

```diff
@@ -0,0 +1,153 @@
+---
+title: C++ persistent containers - array
+author: igchor
+layout: post
+identifier: cpp_array
+---
+
+### Introduction
+
+Until now, our C++ bindings were missing one important component - persistent
+containers. In 1.5 release we have introduced the first one - [pmem::obj::array][cpp_array].
+This container is currently placed in `experimental` namespace and folder. It has
+almost the same functionality as `std::array` from C++11 but takes care of adding
+elements to a transaction and guarantees a fixed layout. API for `pmem::obj::array`
+and `std::array` is the same, except for the following:
+- `pmem::obj::array` defines [range()][cpp_array_range] method (will be described below)
+- `pmem::obj::array` does not mark any non-const function as `noexcept` -
+	elements must be added to a transaction which could result in an exception
+
+If you want to store sequence (which length is known at compile time) of objects
+on persistent memory you should always use `pmem::obj::array`. When using simple
+arrays you must remember to add modified elements to a transaction.
+`pmem::obj::array` does this automatically and makes operations less error prone.
+
+### Usage
+
+Let's start with a simple example:
+
+{% highlight C linenos %}
+#include <libpmemobj++/experimental/array.hpp>
+
+struct data {
+	data() {
+		array = {6, 5, 4, 3, 2, 1};
+	}
+
+	pmem::obj::experimental::array<int, 6> array;
+}
+
+pmem::obj::transaction::run(pop, [&] {
+	ptr = pmem::obj::make_persistent<data>();
+
+	for (auto &e : ptr->array)
+		e++;
+
+	for (auto it = ptr->array.begin(); it != ptr->array.end(); it++)
+		*it += 2;
+
+	for (int i = 0; i < ptr->array.size(); i++)
+		ptr->array[i]--;
+});
+
+{% endhighlight %}
+
+As you can see, you can use `pmem::obj::array` just like an ordinary `std::array`.
+For iterating over it you can use indexing operator, range-based for loops or
+iterators. You can also use `std` algorithms to process the array:
+
+{% highlight C linenos %}
+pmem::obj::transaction::run(pop, [&] {
+	std::sort(ptr->array.begin(), ptr->array.end());
+}
+{% endhighlight %}
+
+If there is an active transaction elements (accessed using any of the above listed
+methods) are snapshotted. In case of iterators returned by begin() and end()
+snapshotting happens during iterator dereferencing. Of course snapshotting is done
+only for elements which can be modified. In case of `const` iterators or `const`
+versions of indexing operator, nothing is added to a transaction. That's why
+it is extremely important to use `const` functions (cbegin(), cend(), etc.)
+whenever possible. It will reduce number of snapshots and can have significant
```

Comment:
`can significantly reduce the performance impact of transactions.`
You usually use "performance impact" in the negative context.